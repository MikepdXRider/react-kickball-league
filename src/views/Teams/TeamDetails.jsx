import React, { useEffect, useState } from 'react'
import { useParams } from 'react-router';
import PlayersList from '../../components/PlayersList/PlayersList.jsx';
import TeamCard from '../../components/TeamCard/TeamCard.jsx';
import { getTeamById } from '../../services/teams.jsx';

export default function TeamDetails() {
    const { id } = useParams();

    const [teamDataObj, setTeamDataObj] = useState({})
    const [isLoading, setIsLoading] = useState(false);

    useEffect(() => {
        setIsLoading(true);
        async function getTeamData() {
            const newTeamDataObj = await getTeamById(id);
            console.log('getTeamById fn returns: ', newTeamDataObj);
            console.log(newTeamDataObj.players);
            setTeamDataObj(newTeamDataObj);
        }
        
        getTeamData();
        setIsLoading(false);
    }, [id])
    
    return (
        <main>
            {
                // ❗ ODD BEHAVIOR ❗
                // THE SITE IS NOT SHOWING 'Loading...'AT ANY POINT!
                isLoading 
                ? <h1>Loading...</h1> 
                : <>
                    <section>
                        <TeamCard teamDataObj={teamDataObj} />
                    </section>
                    <section>
                        {
                            //❗ ODD BEHAVIOR ❗
                            // https://dev.to/samba_code/nested-ternary-statements-in-react-jsx-35kp
                            // replacing teamDataObj.players with teamDataObj.players[0] and attemtping to conditionally render an alternative text breaks the site 
                            // error reads 'TypeError: Cannot read properties of undefined (reading '0')'
                            // If the page is already rendered and you switch out the if conditional below it begins behaving as expected(rending 'No players to display') UNTIL refresh-then every team details page is broken again.
                            // teamDataObj.players[0] ? <PlayersList playersDataArr={teamDataObj.players} /> : <h1>No player data to display</h1>
                            teamDataObj.players && <PlayersList playersDataArr={teamDataObj.players} />
                        }
                    </section>
                </>
            }   
        </main>
    )
}
