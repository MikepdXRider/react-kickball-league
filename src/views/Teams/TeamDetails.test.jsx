import { screen, waitForElementToBeRemoved } from "@testing-library/react";
import { render } from "@testing-library/react";
import { rest } from "msw";
import { setupServer } from "msw/node";
import { MemoryRouter, Route } from "react-router";
import TeamDetails from "./TeamDetails.jsx";

//need to mock out the server for this to work again. 
const mockTeamObj = {
    id: 6,
    created_at: '2021-12-08T20:26:24.408898+00:00',
    name: 'redirect me!',
    city: 'okay',
    state: 'go',
    players: []
  };
  
  const server = setupServer(
    rest.get(
      'https://bmvgpxsbwlwuetqtoczm.supabase.co/rest/v1/teams',
      (req, res, ctx) => {
        return res(ctx.json(mockTeamObj));
      }
    ),
    rest.post(
      'https://bmvgpxsbwlwuetqtoczm.supabase.co/rest/v1/teams',
      (req, res, ctx) => {
        return res(ctx.json([mockTeamObj]));
      }
    )
  );
  
  beforeAll(() => {
    server.listen();
  });
  
  afterAll(() => {
    server.close();
  });

it('renders TeamDetails page w/ data from api', async () => {
    const { container } = render(
    <MemoryRouter initialEntries={[`/teams/1`]}>  
        <Route path='/teams/:id'>
            <TeamDetails />
        </Route>
    </MemoryRouter>
    )
    const loadingText = await screen.findByText(/Delete Team/i);
    // await waitForElementToBeRemoved(() => screen.queryByText(/loading/i))

    expect(loadingText).toBeInTheDocument();
    expect(container).toMatchSnapshot();
})