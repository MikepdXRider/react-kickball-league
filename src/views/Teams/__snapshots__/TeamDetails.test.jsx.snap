// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders TeamDetails page w/ data from api 1`] = `
<div>
  <main>
    <section>
      <article
        style="border: 1px solid black;"
      >
        <article>
          <h3>
            redirect me!
          </h3>
          <h6>
            Date Created: 
            12/8/2021
          </h6>
          <h5>
            okay, go 
          </h5>
        </article>
        <button>
          Edit Team
        </button>
        <button>
          Delete Team
        </button>
      </article>
    </section>
    <section>
      <h1>
        No player data to display
      </h1>
    </section>
  </main>
</div>
`;
